### PR TITLE
Clarify function argument names.

### PR DESCRIPTION
**Pull request by bangerth**
We have three functions that can optionally compute something. Whether they do is controlled by a `bool` argument, and the output is put into a pointer variable. This is duplicative: We should just use whether the pointer is non-`nullptr` to control whether or not to compute the optional thing.

### CODE REVIEW

**Comment by gassmoeller** [comment]
I am not sure the bool is duplicative. This function can be called in three ways:
- In a 'single' non iterative solver scheme no pointer is provided and none should be computed
- In iterative schemes the functions is called in two different ways:
  - a pointer is provided and should be filled
  - a pointer is provided but should not be filled, instead it is used to  compute the relative residual that is returned

There may be better ways to indicate which operation is requested, but I think a single pointer is probably not enough (unless we use `*pointer == 0.0` as a condition to indicate the second case above, but that is fragile, because the residual might just be zero because we are converged.

**Comment by bangerth** [comment]
@gassmoeller Oh, I should have checked here. I think you also figured out what I describe in #5300.

**Comment by bangerth** [comment]
Let's see if this actually works finally. I *think* that I finally understood how this is supposed to work; perhaps I have managed to even express my understanding in code. The PR was initially called "Remove a function argument", which I've come to believe is not actually possible, but perhaps the clarification of the argument names is still useful.

File: source/simulator/solver_schemes.cc [hunk 1]
```diff
@@ -993,10 +987,12 @@ namespace aspect
           particle_world->restore_particles();
 
         const double relative_temperature_residual =
-          assemble_and_solve_temperature(nonlinear_iteration == 0, &initial_temperature_residual);
+          assemble_and_solve_temperature(initial_temperature_residual,
+                                         nonlinear_iteration == 0 ? &initial_temperature_residual : 0);
```
Comment (gassmoeller) [inline source/simulator/solver_schemes.cc:991]:
```suggestion
                                         nonlinear_iteration == 0 ? &initial_temperature_residual : nullptr);
```

File: source/simulator/solver_schemes.cc [hunk 2]
```diff
@@ -993,10 +987,12 @@ namespace aspect
...
 
         const std::vector<double>  relative_composition_residual =
-          assemble_and_solve_composition(nonlinear_iteration == 0, &initial_composition_residual);
+          assemble_and_solve_composition(initial_composition_residual,
+                                         nonlinear_iteration == 0 ? &initial_composition_residual : 0);
```
Comment (gassmoeller) [inline source/simulator/solver_schemes.cc:995]:
```suggestion
                                         nonlinear_iteration == 0 ? &initial_composition_residual : nullptr);
```

**Review by gassmoeller — COMMENTED** [review]
I am not sure I find the new version easier to understand than the old one, but it is not worse either. If you address my comments this is ok to merge.

File: source/simulator/solver_schemes.cc (hunk 1 above)
Comment (bangerth) [reply source/simulator/solver_schemes.cc]: Good point!

**Comment by bangerth** [comment]
Yes, I ended up going around a bit with this PR and ended up in a not vastly different place. Perhaps the biggest difference is to rename the variables. My confusion, as stated in #5300, was mainly that the variable was called "initial_residual", but that it is computed based on a boolean argument that at least in this function we don't know is only ever true *the first time around*. It looks like we're putting the *current* residual into the `initial_residual` variable, and that just seems wrong. I think that the current variable naming is easier to read, though the patch does not otherwise vastly improve over what was there before.

**Review by gassmoeller — APPROVED** [review]

**Pull request merged**